File: Striver A2Z/Step 13 - Binary Trees [Traversals, Medium and Hard Problems]/Lec 1 - Traversals/binary-tree-preorder-traversal.cpp

// https://leetcode.com/problems/binary-tree-preorder-traversal/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    void findPreorder(TreeNode*root, vector<int>&arr){
        if(!root) return;
        arr.push_back(root->val);
        findPreorder(root->left, arr);
        findPreorder(root->right, arr);
    }

    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> preorder;
        if(!root) return preorder;
        findPreorder(root, preorder);
        return preorder;
    }
};

File: Striver A2Z/Step 13 - Binary Trees [Traversals, Medium and Hard Problems]/Lec 1 - Traversals/binary-tree-preorder-traversal.js

// https://leetcode.com/problems/binary-tree-preorder-traversal/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

// const preorder = [];

var solve = (root, preorder) => {
    if(!root) return;
    preorder.push(root.val);
    solve(root.left, preorder);
    solve(root.right, preorder);
    return; 
}

var preorderTraversal = function(root) {
    const preorder = [];
    if(!root) return preorder;
    solve(root, preorder);
    return preorder;
};